### ISSUE TITLE
Switch Case wrong syntax not warned?

### ISSUE THREAD

**Issue by wheelhot** [standard/standard#1157]
Here's an example of what I meant

```JavaScript
    switch(fruits) {
        case fruits = "Banana":
            text = "Banana is good!";
        break;
        case fruits = "Orange":
        text = "I am not a fan of orange.";
        break;
        case fruits = "Apple":
        text = "How you like them apples?";
        break;
        default:
        text = "I have never heard of that fruit...";
    }
```

Notice that the case syntax doesn't follow [official guideline](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/switch)

And it doesn't get any sort of warning in VSCode with StandardJS installed.

The odd thing is, and something I hope someone would be able to answer me is: **Why does this code still works on all browsers except Safari?** And there's an additional interesting thing that happens in Safari as you can see below

---

This is how it works in **Chrome**
![chrome copy](https://user-images.githubusercontent.com/12379413/41957273-f7d03ad0-7a18-11e8-9c66-1cae2b9f5c9e.gif)

---

This is how it works in **Safari**
![safari copy](https://user-images.githubusercontent.com/12379413/41957282-01986b82-7a19-11e8-885b-8722984839a9.gif)

---

And here's the weird part, if I turn on **Safari Web Inspector**, it works, but turn it off and it stops working
![safari webinspector copy](https://user-images.githubusercontent.com/12379413/41957312-17638898-7a19-11e8-9fa2-e29577aa6126.gif)

---

So what's the reason and is there a way to config StandardJS to prevent this from happening?

And here's a [JSFiddle](https://jsfiddle.net/xL742bfe/2/) for those who want to see the code in action yourself. Thanks, [w3 Schools for the sample code](https://www.w3schools.com/jsref/tryit.asp?filename=tryjsref_state_switch1)

**Comment by LinusU** [comment]
> Why does this code still works on all browsers except Safari?

I actually have a hard time telling what the code should do, and I'm not sure which browser is doing the correct thing.

You are actually _assigning_ to `fruits` in each switch case.

I'm guessing that you want something like this:

```diff
    switch(fruits) {
-       case fruits = "Banana":
+       case "Banana":
            text = "Banana is good!";
        break;
-       case fruits = "Orange":
+       case "Orange":
        text = "I am not a fan of orange.";
        break;
-       case fruits = "Apple":
+       case "Apple":
        text = "How you like them apples?";
        break;
        default:
        text = "I have never heard of that fruit...";
    }
```

-----

That being said, I'm actually surprised that Standard doesn't warn about this, and it would be very nice to fix that.

**Comment by LinusU** [comment]
Just tested on the "Try standard" feature of the site, the following passes without errors:

```js
var fruit = 'apple'

switch (fruit) {
  case fruit = 'apple':
    console.log('it was apple')
    break
  case fruit = 'banana':
    console.log('it was banana')
    break
  default:
    console.log('i don\'t know that fruit')
}
```

**Comment by wheelhot** [comment]
> I actually have a hard time telling what the code should do, and I'm not sure which browser is doing the correct thing.

Okay, let me rephrase my points:
1. The above code with `case fruits = "Banana":` works fine on all browsers except Safari
2. You can test it by typing the case string in the text box in both Safari and Chrome and you'll notice that it doesn't work in Safari
3. Turn on Safari Web Inspector and it'll work again

So my question was:
1. Why does this wrong syntax, somehow work on all browsers except Safari? (when it should not work at all)
2. why doesn't StandardJS warn that I used the wrong syntax format (hence avoiding all this issue, to begin with)

> You are actually assigning to fruits in each switch case.
Yeah, I realised that but when I code I didn't realise that mistake and no errors on Chrome but when I tested in Safari, things didn't work and turning on Safari Web Inspector makes the code working again, so ended up spending more time figuring out what went wrong only to have someone point out that I got my syntax wrong. So I figure it's best to share my findings and see what people think about it. I know it's trivial, but I'm curious on why

> I'm guessing that you want something like this:
Yes :)

**Comment by LinusU** [comment]
> Why does this wrong syntax

It's not technically the wrong syntax. You are allowed to put any expression in a case statement, and assignment is a valid expression.

A similar trick can be used to simulate pattern matching:

```js
let bugs = 99

switch (true) {
  case (bugs < 10):
    console.log('Pretty good code')
    break
  case (bugs < 50):
    console.log('Code review, anyone?')
    break
  case (bugs < 100):
    console.log('🙈')
    break
  default:
    console.log('I don\'t even')
}
```

> works fine on all browsers except Safari

I'm still not convinced which browser is "correct".

You are seeing different behavior in Safari or Chrome because they are evaluating the case expressions at a different point in time. This is potentially defined in the ECMAScript language spec...
